### PR TITLE
#614 add casbin for authorization

### PR DESCRIPTION
**Pull request by suekto-andreas**
### Enhancement #614 
Add into the middleware: [casbin](https://github.com/casbin/casbin)

Design to be review are
1. the arguments to be pass to NewEnforcer
2. is `auth` directory the right place to put casbin
any further feedback please help to add, thanks !

Disclaimer -- testdata are actually taken from casbin repository test

### CODE REVIEW

**Comment by suekto-andreas** [comment]
Hi @peterbourgon, the tests failed on metrics/pcp - TestInconsistentLabelCardinality, it should be unrelated with this PR (auth/casbin)

can help to advise how should we continue ?

**Comment by peterbourgon** [comment]
Oh cool! Ignore the error, I'll fix it up.

Primary review comment right now, please fix up all the doc comments to be valid English sentences with capitalization and punctuation, word-wrapped at 80 columns.

File: auth/casbin/middleware.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+// while also checking the authorization for the corresponding subject, object, and action
+func NewEnforcer(subject string, object interface{}, action string) endpoint.Middleware {
+	return func(next endpoint.Endpoint) endpoint.Endpoint {
+		return func(ctx context.Context, request interface{}) (response interface{}, err error) {
+			casbinModel := ctx.Value(CasbinModelContextKey)
+			casbinPolicy := ctx.Value(CasbinPolicyContextKey)
+
+			enforcer := stdcasbin.NewEnforcer(casbinModel, casbinPolicy)
+			ctx = context.WithValue(ctx, CasbinEnforcerContextKey, enforcer)
+			if enforcer.Enforce(subject, object, action) == false {
```
Comment (peterbourgon) [inline auth/casbin/middleware.go]:
```suggestion
			if !enforcer.Enforce(subject, object, action) {
```

File: auth/casbin/middleware.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+// NewEnforcer installs casbin enforcer into the context
+// while also checking the authorization for the corresponding subject, object, and action
+func NewEnforcer(subject string, object interface{}, action string) endpoint.Middleware {
+	return func(next endpoint.Endpoint) endpoint.Endpoint {
+		return func(ctx context.Context, request interface{}) (response interface{}, err error) {
+			casbinModel := ctx.Value(CasbinModelContextKey)
+			casbinPolicy := ctx.Value(CasbinPolicyContextKey)
+
+			enforcer := stdcasbin.NewEnforcer(casbinModel, casbinPolicy)
+			ctx = context.WithValue(ctx, CasbinEnforcerContextKey, enforcer)
```
Comment (peterbourgon) [inline auth/casbin/middleware.go]: Can this be moved to after the Enforce check? e.g. delete it here, and call next with

```go
return next(context.WithValue(ctx, CasbinEnforcerContextKey, enforcer), request)
```

**Comment by peterbourgon** [comment]
If you rebase the test errors should go away.

File: auth/casbin/middleware.go (hunk 1 above)
Comment (suekto-andreas) [reply auth/casbin/middleware.go]: thanks for the suggestion ! shall apply it together with the docs & word wrap

**Comment by suekto-andreas** [comment]
thanks @peterbourgon, the pcp is good now but the circleci got err on sd/etcdv3 integration test part

File: auth/casbin/middleware.go (hunk 2 above)
Comment (peterbourgon) [reply auth/casbin/middleware.go]: Ah, it's reassigned, I understand, I think 👍

File: auth/casbin/middleware.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	stdcasbin "github.com/casbin/casbin"
+	"github.com/go-kit/kit/endpoint"
+)
+
+type contextKey string
+
+const (
+	// CasbinModelContextKey holds the key to store the access control model
+	// in context, it can be a path to configuration file or a casbin/model
+	// Model
```
Comment (peterbourgon) [inline auth/casbin/middleware.go]: Still need punctuation etc. on all doc comments like this one.

Comment (suekto-andreas) [reply auth/casbin/middleware.go]: submitted and update, would it be like that ? sorry kinda new into this.

https://github.com/go-kit/kit/pull/787/commits/bf51e5ee889a96124f7d81177a53b9bb5ce56d64

**Comment by peterbourgon** [comment]
Thanks!

**Pull request merged**